include/cetl/span.h

```diff
@@ -1 +1,772 @@
+               Count != cetl::dynamic_extent
+                   ? Count
+                   : extent - Offset > {data_ + Offset, Count == cetl::dynamic_extent ? size() - Offset : Count};
+    }
+
+    /// Create a new span `Offset` elements from the start of the current span and for either `Count` elements or,
+    /// if Count is `cetl::dynamic_extent`, for the remaining size of the span (i.e. `size() - Offset`).
+    /// The behavior of this method is undefined where `Count` is greater-than the size of this span or if
+    /// `Count` is not cetl::dynamic_extent but `Count` is greater-than `size() - Offset`.
+    /// @param Offset  Number of elements from the start of the current span for the subspan.
+    /// @param Count   The number of elements from the Offset to include in the subspan. If this value is
+    ///                `cetl::dynamic_extent` then the Count is `size() - Offset`.
+    /// @return A new span with a dynamic extent.
+    /// @see std::span::subspan()
+    constexpr span<element_type, cetl::dynamic_extent> subspan(size_type Offset,
+                                                               size_type Count = cetl::dynamic_extent) const
+    {
+        CETL_DEBUG_ASSERT(Offset <= extent, "subspan Offsets > size() are ill-formed.");
+        if (Count == cetl::dynamic_extent)
+        {
+            return {data_ + Offset, extent - Offset};
+        }
+        else
+        {
+            CETL_DEBUG_ASSERT(Count <= (extent - Offset), "subspan Count argument is ill-formed");
+            return {data_ + Offset, Count};
+        }
+    }
+
+    ///@}
+private:
+    pointer data_;
+};
+
+/// Specialization of span where the extent is dynamic.
+template <typename T>
+class span<T, cetl::dynamic_extent>
+{
+public:
+    // (Groupings per https://en.cppreference.com/w/cpp/container/span)
+    // +----------------------------------------------------------------------+
+    // | Member Constants
+    // +----------------------------------------------------------------------+
+    /// The value of Extent for this template instantiation.
+    static constexpr std::size_t extent = cetl::dynamic_extent;
+
+    // +----------------------------------------------------------------------+
+    // | Member Types
+    // +----------------------------------------------------------------------+
+    using element_type     = T;                                 //!< The element type over which the span operates
+    using value_type       = typename std::remove_cv<T>::type;  //!< The non-cv type of elements
+    using pointer          = T*;                                //!< The element pointer type
+    using const_pointer    = const T*;                          //!< A constant pointer to the element type
+    using reference        = T&;                                //!< A reference type to the element
+    using const_reference  = const T&;                          //!< A constant reference type to the element
```

Comment:
```suggestion
    using reference_const  = T const&;                          //!< A reference to the constant element type
```

Reply:
Why this change? This is taken from the C++ standard directly:
![image](https://user-images.githubusercontent.com/19180861/223512385-96470b9d-5824-468a-bb63-3f01506ee17f.png)

Reply:
Even the standard doesn't logically make sense as the reference isn't constant, the object it references is.